tensorflow_addons/losses/contrastive.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements contrastive loss."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow_addons.utils import keras_utils
+
+
+@keras_utils.register_keras_custom_object
+@tf.function
+def contrastive_loss(y_true, y_pred, margin=1.0):
+    """Computes the contrastive loss between `y_true` and `y_pred`.
+
+    This loss encourages the embedding to be close to each other for
+    the samples of the same label and the embedding to be far apart at least
+    by the margin constant for the samples of different labels.
+
+    The euclidean distances `y_pred` between two embedding matrices
+    `a` and `b` with shape [batch_size, hidden_size] can be computed
```

Comment:
two l2-normalized embedding matrices `a` and `b` ?

Reply:
In [keras example](https://github.com/keras-team/keras/blob/master/examples/mnist_siamese.py), the two embedding seems not to be 12-normalized. I also check original paper and find nothing about L2 normalization.

Reply:
I find it in the document of original implementation:
> Embeddings should be l2 normalized.

https://github.com/tensorflow/tensorflow/blob/3cd3277b81406c9de1a02b401b1fd0a7ad72f0b3/tensorflow/contrib/losses/python/metric_learning/metric_loss_ops.py#L99

but it's test case doesn't conform the restriction:
https://github.com/tensorflow/tensorflow/blob/3cd3277b81406c9de1a02b401b1fd0a7ad72f0b3/tensorflow/contrib/losses/python/metric_learning/metric_loss_ops_test.py#L69

And in the original paper
> No assumption is made about the function G_W except that it is differentiable with respect to W

So I think we don't need to require all embedding l2-normalized.